experimental/Schemes/elliptic_surface.jl

```diff
@@ -0,0 +1,1116 @@
+  #Todo basis
+  return lll(mwl)
+end
+
+@doc raw"""
+    mordell_weil_torsion(S::EllipticSurface) ->
+
+Return generators for the torsion part of the Mordell-Weil group
+of the generic fiber of ``S``.
+"""
+@attr function mordell_weil_torsion(S::EllipticSurface)
+  E = generic_fiber(S)
+  O = E([0,1,0])
+  N = trivial_lattice(S)[2]
+  tors = EllCrvPt[]
+  d = det(N)
+  for p in prime_divisors(d)
+    if valuation(d, p) == 1
+      continue
+    end
+    r = 1
+    i = 0
+    while true
+      i = i+1
+      @vprint :EllipticSurface 2 "computing $(p^i)-torsion"
+      global dp = division_points(O, p^i)
+      if length(dp) == r
+        break
+      end
+      r = length(dp)
+    end
+    for pt in dp
+      if pt != O
+        push!(tors, pt)
+      end
+    end
+  end
+  return tors
+end
+
+@doc raw"""
+
+#Examples
+```jldoctest
+julia> Qt, t = polynomial_ring(QQ, :t);
+
+julia> Qtf = fraction_field(Qt);
+
+julia> E = EllipticCurve(Qtf, [0,0,0,0,t^5*(t-1)^2]);
+
+julia> Base.show(stdout, X3)
+Elliptic surface with generic fiber -x^3 + y^2 - t^7 + 2*t^6 - t^5
+
+```
+"""
+function Base.show(io::IO, S::EllipticSurface)
+  io = pretty(io)
+  if get(io, :supercompact, false)
+    print(io, "Elliptic surface")
+  else
+    E = generic_fiber(S)
+    print(io, "Elliptic surface with generic fiber ", equation(E))
+  end
+end
+
+@doc raw"""
+#Examples
+```jldoctest
+julia> Qt, t = polynomial_ring(QQ, :t);
+
+julia> Qtf = fraction_field(Qt);
+
+julia> E = EllipticCurve(Qtf, [0,0,0,0,t^5*(t-1)^2]);
+
+julia> X3 = elliptic_surface(E, 2)
+Elliptic surface
+  with generic fiber
+    elliptic curve with equation
+        y^2 = x^3 + t^7 - 2*t^6 + t^5
+
+```
+"""
+function Base.show(io::IO, ::MIME"text/plain", S::EllipticSurface)
+  io = pretty(io)
+  println(io, "Elliptic surface")
+  println(io, Indent(), "with generic fiber")
+  print(io, Indent(), Lowercase(), generic_fiber(S), Dedent())
+  if isdefined(S, :Y)
+    println(io, "")
+    println(io, "with relatively minimal model")
+    print(io, Indent(), Lowercase(), relatively_minimal_model(S)[1], Dedent())
+  end
+  print(io, Dedent())
+end
+
+@doc raw"""
+    weierstrass_model(X::EllipticSurface) -> CoveredScheme, CoveredClosedEmbedding
+
+Return the weierstrass model ``S`` of ``X`` and the inclusion
+
+\[S \subseteq \mathbb{P}( \mathcal{O}_{\mathbb{P}^1}(-2s) \oplus \mathcal{O}_{\mathbb{P}^1}(-3s) \oplus \mathcal{O}_{\mathbb{P}^1})\]
+"""
+function weierstrass_model(X::EllipticSurface)
+  if isdefined(X, :Weierstrassmodel)
+    return X.Weierstrassmodel, X.inc_Weierstrass
+  end
+
+  s = X.bundle_number
+  E = generic_fiber(X)
+
+  kt = base_ring(base_field(E))
+  k = coefficient_ring(kt)
+  delta = factor(discriminant(E), kt).fac
+  reducible_singular_fibers = [p for p in keys(delta) if delta[p]>1]
+
+  IP1 = projective_space(k, 1)
+  c = standard_covering(IP1)
+  # rename the variables on the affine charts
+  # to a more readable version
+  OO(c[1]).S = [:t]
+  OO(c[2]).S = [:s]
```

Comment:
Let us discuss some reasonable entry points for this. You shouldn't have to do it this way.